src/rust/engine/process_execution/src/nailgun/nailgun_pool.rs

```diff
@@ -262,64 +300,113 @@ pub struct NailgunProcess {
   pub fingerprint: NailgunProcessFingerprint,
   pub build_id: String,
   pub port: Port,
-  pub handle: Arc<Mutex<std::process::Child>>,
+  pub handle: Arc<parking_lot::Mutex<std::process::Child>>,
 }
 
-fn read_port(child: &mut std::process::Child) -> Result<Port, String> {
-  let stdout = child
-    .stdout
-    .as_mut()
-    .ok_or_else(|| "No Stdout found!".to_string());
-  stdout.and_then(|stdout| {
-    let reader = io::BufReader::new(stdout);
-    let line = reader
-      .lines()
-      .next()
-      .ok_or("There is no line ready in the child's output")?
-      .map_err(|err| format!("{}", err))?;
-    let port = &NAILGUN_PORT_REGEX
-      .captures_iter(&line)
-      .next()
-      .ok_or("Output for nailgun server didn't match the regex!")?[1];
-    port
-      .parse::<Port>()
-      .map_err(|e| format!("Error parsing port {}! {}", &port, e))
+fn read_port(stdout: PathBuf) -> BoxFuture<Port, String> {
+  future::loop_fn(MAX_PROCESS_OUTPUT_POLLS, move |mut loops| {
+    let stdout = stdout.clone();
+    let wait_period = Duration::from_millis(100);
+    sleep(wait_period)
+      .map_err(|e| format!("sleep while waiting for nailgun stdout failed: {:?}", e))
+      .and_then(move |_| {
+        fs::File::open(stdout.clone()).map_err(move |_| format!("Could not open file {:?}", stdout))
+      })
+      .and_then(|log| {
+        lines(BufReader::new(log))
+          .take(1)
+          .into_future()
+          .map_err(|(err, _s)| format!("Error getting file line in read port {}", err))
+      })
+      .and_then(move |(line, _)| {
+        if let Some(s) = line {
+          debug!("Nailgun process startup output is: `{:?}`", s);
+          Ok(Loop::Break(Ok(s)))
+        } else {
+          loops -= 1;
+          if loops == 0 {
```

Comment:
Let's phrase this timeout in terms of elapsed time, not number of iterations.

Reply:
Will do!